core/io/close_spec.rb

```diff
@@ -45,6 +45,20 @@
       @io.close.should be_nil
     end
   end
+
+  ruby_version_is '2.5' do
+    it 'raises an IOError with a clear message' do
+      read_io, write_io = IO.pipe
+      thread = Thread.new do
+        ->{read_io.read}.should raise_error(IOError, 'stream closed in another thread')
+      end
+
+      Thread.pass until thread.stop?
```

Comment:
Could you add an extra local variable to track we are just before the `read` ?
Otherwise this might not wait long enough if some Ruby implementation does a blocking action (setting `Thread#status` to `"sleep"`) before reaching the `read`.
Something like:
```ruby
going_to_read = false

  # In the lambda
  going_to_read = true
  read_io.read

Thread.pass until going_to_read && thread.stop?
```

Reply:
In case the `Thread#status` is `"sleep"` thread.stop? will still return true then do we still need this `going_to_read` check? I tried --
```
2.5.0 :016 > t = Thread.new { sleep }
 => #<Thread:0x000055797c7fd780@(irb):16 run> 
2.5.0 :017 > t.status
 => "sleep" 
2.5.0 :018 > t.stop?
 => true 
2.5.0 :019 > p 'dead' if t.stop?
"dead"
2.5.0 :020 > p 'alive' unless t.stop?
 => nil 
```
Can you please explain? Maybe I didn't get `this might not wait long enough if some Ruby implementation does a blocking action`

Reply:
For instance with something like:
```ruby
t = Thread.new do
  sleep 0.1
  read_io.read
end

Thread.pass until thread.stop?
```
This might or might not wait until we are in `read_io.read`.

This is a counter-example, but it is not impossible that a blocking action happens in `should`, 
 `raise_error` or even in `Thread.new`.
I know TruffleRuby used to have both `Thread.new` and regexp matching considered as blocking actions (setting the Thread#status to `"sleep"` and therefore `thread.stop?` would return `true`). These are compatibility bugs and they got fixed.
But it seems best to avoid relying on the fact no blocking action is performed between in `Thread.new`, `should` and `raise_error` before we reach the `read_io.read`.

Admittedly I might be too afraid of such things, but I think a bit of extra safety doesn't hurt :smiley:

Reply:
I tried ^ and got to know how its a blocking action. Thank you for explaining. I have updated the PR.